dask/dataframe/core.py

```diff
@@ -2630,12 +2630,21 @@ def _is_single_last_div(x):
 
 def repartition_npartitions(df, npartitions):
     """ Repartition dataframe to a smaller number of partitions """
-    npartitions = min(npartitions, df.npartitions)
-    k = int(math.ceil(df.npartitions / npartitions))
-    divisions = df.divisions[::k]
-    if len(divisions) <= npartitions:
-        divisions = divisions + (df.divisions[-1],)
-    return df.repartition(divisions=divisions)
+    size = df.npartitions / npartitions
+    L = [int(i * size) for i in range(npartitions + 1)]
+    name = 'repartition-%d-%s' % (npartitions, df._name)
+    dsk = dict(
+        (
+            (name, i),
+            (pd.concat, [(df._name, j) for j in range(L[i], L[i + 1])])
+        ) for i in range(npartitions)
+    )
```

Comment:
I haven't heard of list/dict comprehensions in JS...

I would rather do dict comprehension

```python
dsk = {(name, i): (pd.concat, [...]) for ...}
```

instead of `dict(iterator)`, but for some reason this style is avoided in Dask. Do you have an insight?